wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java

```diff
@@ -487,6 +487,29 @@ public void createSnapshot(String workspaceId) throws NotFoundException, ServerE
         return machineManager.getSnapshots(sessionUser().getId(), workspaceId);
     }
 
+    /**
+     * Adds machine into workspace runtime.
+     *
+     * @param machine
+     *         machine to add to specified runtime
+     * @throws ConflictException
+     *         if given machine already exists in specified runtime
+     *         or if machine is dev-machine
+     * @throws NotFoundException
+     *         if machine's workspace hasn't runtime
+     */
+    public void injectMachine(MachineImpl machine) throws ConflictException, NotFoundException, ServerException {
+        if (machine.getConfig().isDev()) {
+            throw new ConflictException("Cannot add another dev-machine");
+        }
+        String workspaceId = machine.getWorkspaceId();
+        if (runtimes.get(workspaceId).getRuntimeStatus() != WorkspaceStatus.RUNNING) {
```

Comment:
You can't guarantee that it is true unless you are in the lock, so this is not the right place for such check, move it to the `WorkspaceRuntimes` method.